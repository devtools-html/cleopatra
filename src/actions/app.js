/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { getSelectedTab, getDataSource } from '../selectors/url-state';
import { getTrackThreadHeights } from '../selectors/app';
import { sendAnalytics } from '../utils/analytics';
import type { ThreadIndex } from '../types/profile';
import type { CssPixels } from '../types/units';
import type { Action, ThunkAction } from '../types/store';
import type { TabSlug } from '../app-logic/tabs-handling';
import type { UrlState } from '../types/state';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export const reportTrackThreadHeight = (
  threadIndex: ThreadIndex,
  height: CssPixels
): ThunkAction<void> => (dispatch, getState) => {
  const trackThreadHeights = getTrackThreadHeights(getState());
  const previousHeight = trackThreadHeights[threadIndex];
  if (previousHeight !== height) {
    // Guard against unnecessary dispatches. This could happen frequently.
    dispatch({
      type: 'UPDATE_TRACK_THREAD_HEIGHT',
      height,
      threadIndex,
    });
  }
};

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}
