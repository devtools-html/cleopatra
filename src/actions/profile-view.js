/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  ProfileSelection, CallTreeFilter, ImplementationFilter,
} from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(({
      type: 'HIDE_THREAD',
      threadIndex,
      threadOrder,
      hiddenThreads,
    }: Action));
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
