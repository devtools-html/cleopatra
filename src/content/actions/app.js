// @flow
import type { Action } from './types';

export function changeSelectedTab(selectedTab: string): Action {
  return {
    type: 'CHANGE_SELECTED_TAB',
    selectedTab,
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeTabOrder(tabOrder: number[]): Action {
  return {
    type: 'CHANGE_TAB_ORDER',
    tabOrder,
  };
}
