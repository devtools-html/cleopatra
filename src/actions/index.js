/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
