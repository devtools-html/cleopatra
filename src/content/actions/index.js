// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
