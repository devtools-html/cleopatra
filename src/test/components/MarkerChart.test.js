/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { Provider } from 'react-redux';

// This module is mocked.
import copy from 'copy-to-clipboard';

import { changeMarkersSearchString } from '../../actions/profile-view';
import {
  TIMELINE_MARGIN_LEFT,
  TIMELINE_MARGIN_RIGHT,
} from '../../app-logic/constants';
import MarkerChart from '../../components/marker-chart';
import MarkerContextMenu from '../../components/shared/MarkerContextMenu';
import { changeSelectedTab } from '../../actions/app';
import { ensureExists } from '../../utils/flow';

import mockCanvasContext from '../fixtures/mocks/canvas-context';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithMarkers } from '../fixtures/profiles/processed-profile';
import {
  getBoundingBox,
  getMouseEvent,
  addRootOverlayElement,
  removeRootOverlayElement,
} from '../fixtures/utils';
import mockRaf from '../fixtures/mocks/request-animation-frame';

import type { CssPixels } from '../../types/units';

const MARKERS = [
  ['Marker A', 0, { startTime: 0, endTime: 10 }],
  ['Marker A', 0, { startTime: 0, endTime: 10 }],
  ['Marker A', 11, { startTime: 11, endTime: 15 }],
  [
    'Very very very very very very Very very very very very very Very very very very very very Very very very very very very Very very very very very very long Marker D',
    6,
    { startTime: 5, endTime: 15 },
  ],
  ['Dot marker E', 4, { startTime: 4, endTime: 4 }],
  ['Non-interval marker F without data', 7, null],
  [
    'Marker G type DOMEvent',
    5,
    {
      type: 'tracing',
      category: 'DOMEvent',
      eventType: 'click',
      interval: 'start',
      phase: 2,
    },
  ],
  [
    'Marker G type DOMEvent',
    10,
    {
      type: 'tracing',
      category: 'DOMEvent',
      eventType: 'click',
      interval: 'end',
      phase: 2,
    },
  ],
  [
    'Marker H with no start',
    3,
    {
      type: 'tracing',
      category: 'Paint',
      interval: 'end',
    },
  ],
  [
    'Marker H with no end',
    9,
    {
      type: 'tracing',
      category: 'Paint',
      interval: 'start',
    },
  ],
];

function setupWithProfile(profile) {
  const flushRafCalls = mockRaf();
  const ctx = mockCanvasContext();
  jest
    .spyOn(HTMLCanvasElement.prototype, 'getContext')
    .mockImplementation(() => ctx);

  // Ideally we'd want this only on the Canvas and on ChartViewport, but this is
  // a lot easier to mock this everywhere.
  jest
    .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
    .mockImplementation(() =>
      getBoundingBox(200 + TIMELINE_MARGIN_LEFT + TIMELINE_MARGIN_RIGHT, 300)
    );

  const store = storeWithProfile(profile);
  store.dispatch(changeSelectedTab('marker-chart'));

  const renderResult = render(
    <Provider store={store}>
      <>
        <MarkerContextMenu />
        <MarkerChart />
      </>
    </Provider>
  );

  const { container } = renderResult;

  function fireMouseEvent(eventName, options) {
    fireEvent(
      ensureExists(
        container.querySelector('canvas'),
        `Couldn't find the canvas element`
      ),
      getMouseEvent(eventName, options)
    );
  }

  return {
    ...renderResult,
    ...store,
    flushRafCalls,
    flushDrawLog: () => ctx.__flushDrawLog(),
    fireMouseEvent,
  };
}

describe('MarkerChart', function() {
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  it('renders the normal marker chart and matches the snapshot', () => {
    window.devicePixelRatio = 2;

    const profile = getProfileWithMarkers([...MARKERS]);
    const {
      container,
      flushRafCalls,
      dispatch,
      flushDrawLog,
    } = setupWithProfile(profile);

    dispatch(changeSelectedTab('marker-chart'));
    flushRafCalls();

    const drawCalls = flushDrawLog();
    expect(container.firstChild).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();

    delete window.devicePixelRatio;
  });

  it('does not render several dot markers on the same pixel', () => {
    window.devicePixelRatio = 1;
    const markers = [
      // 'Marker first' and 'Marker last' define our range.
      ['Marker first', 0, null],
      // Then 2 very close dot markers with the same name. They shouldn't be
      // drawn both together.
      ['Marker A', 5000, null],
      ['Marker A', 5001, null],
      // This is a longer marker, it should always be drawn even if it starts at
      // the same location as a dot marker
      ['Marker A', 5001, { startTime: 5001, endTime: 7000 }],
      [
        'Marker last',
        15000,
        null,
      ] /* add a marker that's quite far away to have a big range */,
    ];

    const profile = getProfileWithMarkers(markers);
    const { flushRafCalls, flushDrawLog } = setupWithProfile(profile);
    flushRafCalls();

    const drawCalls = flushDrawLog();

    // Check that we have 3 arc operations (first marker, one of the 2 dot
    // markers in the middle, and last marker)
    const arcOperations = drawCalls.filter(
      ([operation]) => operation === 'arc'
    );
    expect(arcOperations).toHaveLength(3);

    // Check that all X values are different
    const arcOperationsX = new Set(arcOperations.map(([, x]) => Math.round(x)));
    expect(arcOperationsX.size).toBe(3);

    // Check that we have a fillRect operation for the longer marker.
    // We filter on the height to get only 1 relevant fillRect operation per marker
    const fillRectOperations = drawCalls.filter(
      ([operation, , , , height]) =>
        operation === 'fillRect' && height > 1 && height < 16
    );
    expect(fillRectOperations).toHaveLength(1);

    delete window.devicePixelRatio;
  });

  it('renders the hoveredItem markers properly', () => {
    window.devicePixelRatio = 1;

    const profile = getProfileWithMarkers(MARKERS);
    const {
      flushRafCalls,
      dispatch,
      flushDrawLog,
      fireMouseEvent,
    } = setupWithProfile(profile);

    dispatch(changeSelectedTab('marker-chart'));
    flushRafCalls();
    flushDrawLog();

    // No tooltip displayed yet
    expect(document.querySelector('.tooltip')).toBeFalsy();

    // Move the mouse on top of an item.
    fireMouseEvent('mousemove', {
      offsetX: 200,
      offsetY: 5,
      pageX: 200,
      pageY: 5,
    });

    flushRafCalls();

    const drawCalls = flushDrawLog();
    expect(drawCalls).toMatchSnapshot();

    // The tooltip should be displayed
    expect(document.querySelector('.tooltip')).toMatchSnapshot();
  });

  describe('displays context menus', () => {
    beforeEach(() => {
      // Always use fake timers when dealing with context menus.
      jest.useFakeTimers();
    });

    function setup() {
      const profile = getProfileWithMarkers(MARKERS);
      const setupResult = setupWithProfile(profile);
      const {
        flushRafCalls,
        dispatch,
        flushDrawLog,
        fireMouseEvent,
        container,
        getByText,
      } = setupResult;

      dispatch(changeSelectedTab('marker-chart'));
      flushRafCalls();
      flushDrawLog();

      function getPositioningOptions({ x, y }) {
        // These positioning options will be sent to all our mouse events. Note
        // that the values aren't really consistent, especially offsetY and
        // pageY shouldn't be the same, but in the context of our test this will
        // be good enough.
        // pageX/Y values control the position of the tooltip so it's not super
        // important.
        // offsetX/Y are more important as they're used to find which node is
        // actually clicked.
        const positioningOptions = {
          offsetX: x,
          offsetY: y,
          clientX: x,
          clientY: y,
          pageX: x,
          pageY: y,
        };

        return positioningOptions;
      }

      // Note to a future developer: the x/y values can be derived from the
      // array returned by flushDrawLog().
      function rightClick(where: { x: CssPixels, y: CssPixels }) {
        const positioningOptions = getPositioningOptions(where);
        const clickOptions = {
          ...positioningOptions,
          button: 2,
          buttons: 2,
        };

        // Because different components listen to different events, we trigger
        // all the right events, to be as close as possible to the real stuff.
        fireMouseEvent('mousemove', positioningOptions);
        fireMouseEvent('mousedown', clickOptions);
        fireMouseEvent('mouseup', clickOptions);
        fireMouseEvent('contextmenu', clickOptions);
        flushRafCalls();
      }

      function mouseOver(where: { x: CssPixels, y: CssPixels }) {
        const positioningOptions = getPositioningOptions(where);
        fireMouseEvent('mousemove', positioningOptions);
        flushRafCalls();
      }

      function clickOnMenuItem(stringOrRegexp) {
        fireEvent.click(getByText(stringOrRegexp));
      }

      const getContextMenu = () =>
        ensureExists(
          container.querySelector('.react-contextmenu'),
          `Couldn't find the context menu.`
        );

      return {
        ...setupResult,
        rightClick,
        mouseOver,
        getContextMenu,
        clickOnMenuItem,
      };
    }

    it('when right clicking on a marker', () => {
      const { rightClick, clickOnMenuItem, getContextMenu } = setup();

      // The "Marker A" marker is drawn from 150,1 to 275,13.
      rightClick({ x: 200, y: 5 });
      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      clickOnMenuItem('Copy');
      expect(copy).toHaveBeenLastCalledWith('Marker A');
      expect(getContextMenu()).not.toHaveClass('react-contextmenu--visible');

      jest.runAllTimers();
      expect(document.querySelector('react-contextmenu')).toBeFalsy();
    });

    it('when right clicking on markers in a sequence', () => {
      const { rightClick, clickOnMenuItem, getContextMenu } = setup();

      // The "Marker A" marker is drawn from 150,1 to 275,13.
      rightClick({ x: 200, y: 5 });
      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      // The "click" DOMEvent marker is drawn from 213,82 to 275.5,93.
      rightClick({ x: 220, y: 90 });
      jest.runAllTimers();

      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');
      clickOnMenuItem('Copy');
      expect(copy).toHaveBeenLastCalledWith('click');
    });

    it('and still highlights other markers when hovering them', () => {
      const { rightClick, mouseOver, flushDrawLog, getContextMenu } = setup();

      // The "Marker A" marker is drawn from 150,1 to 275,13.
      rightClick({ x: 200, y: 5 });
      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      flushDrawLog();
      // The "click" DOMEvent marker is drawn from 213,82 to 275.5,93.
      mouseOver({ x: 220, y: 90 });

      // Expect that we have 2 markers drawn with this color.
      const drawCalls = flushDrawLog();
      const callsWithHighlightColor = drawCalls.filter(
        ([, argument]) => argument === 'Highlight'
      );
      expect(callsWithHighlightColor).toHaveLength(2);
    });
  });

  describe('with search strings', function() {
    function getFillTextCalls(drawCalls) {
      return drawCalls
        .filter(([methodName]) => methodName === 'fillText')
        .map(([_, text]) => text);
    }

    const searchString = 'Dot marker E';

    it('renders lots of markers initially', function() {
      const profile = getProfileWithMarkers(MARKERS);
      const { flushRafCalls, flushDrawLog } = setupWithProfile(profile);

      flushRafCalls();
      const text = getFillTextCalls(flushDrawLog());
      expect(text.length).toBeGreaterThan(1);
      // Check that our test search string is in here:
      expect(text.filter(t => t === searchString).length).toBe(1);
    });

    it('renders only the marker that was searched for', function() {
      const profile = getProfileWithMarkers(MARKERS);
      const { flushRafCalls, dispatch, flushDrawLog } = setupWithProfile(
        profile
      );

      // Flush out any existing draw calls.
      flushRafCalls();
      flushDrawLog();

      // Update the chart with a search string.
      dispatch(changeMarkersSearchString(searchString));
      flushRafCalls();

      const text = getFillTextCalls(flushDrawLog());
      expect(text.length).toBe(1);
      expect(text[0]).toBe(searchString);
    });
  });

  describe('EmptyReasons', () => {
    it('shows a reason when a profile has no markers', () => {
      const profile = getProfileWithMarkers([]);
      const { dispatch, container } = setupWithProfile(profile);

      dispatch(changeSelectedTab('marker-chart'));
      expect(container.querySelector('.EmptyReasons')).toMatchSnapshot();
    });

    it("shows a reason when a profile's markers have been filtered out", () => {
      const profile = getProfileWithMarkers(MARKERS);
      const { dispatch, container } = setupWithProfile(profile);

      dispatch(changeSelectedTab('marker-chart'));
      dispatch(changeMarkersSearchString('MATCH_NOTHING'));
      expect(container.querySelector('.EmptyReasons')).toMatchSnapshot();
    });
  });
});
