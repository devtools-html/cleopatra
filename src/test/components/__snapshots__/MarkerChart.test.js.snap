// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkerChart EmptyReasons shows a reason when a profile has no markers 1`] = `
<div
  className="EmptyReasons"
>
  <h3>
    The 
    marker chart
     is empty for “
    Empty
    ”
  </h3>
  <p>
    This thread contains no markers for this chart.
  </p>
</div>
`;

exports[`MarkerChart EmptyReasons shows a reason when a profile's markers have been filtered out 1`] = `
<div
  className="EmptyReasons"
>
  <h3>
    The 
    marker chart
     is empty for “
    Empty
    ”
  </h3>
  <p>
    All markers were filtered out by the current selection or search term.
  </p>
</div>
`;

exports[`MarkerChart renders the hoveredItem markers properly 1`] = `
Array [
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    365,
    300,
  ],
  Array [
    "set lineWidth",
    1,
  ],
  Array [
    "set fillStyle",
    "Highlight",
  ],
  Array [
    "fillRect",
    151,
    1,
    248,
    1,
  ],
  Array [
    "fillRect",
    150,
    2,
    250,
    11,
  ],
  Array [
    "fillRect",
    151,
    13,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    151,
    17,
    248,
    1,
  ],
  Array [
    "fillRect",
    150,
    18,
    250,
    11,
  ],
  Array [
    "fillRect",
    151,
    29,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "arc",
    255,
    39.5,
    3.75,
    0,
    6.283185307179586,
  ],
  Array [
    "fill",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    276,
    49,
    248,
    1,
  ],
  Array [
    "fillRect",
    275,
    50,
    250,
    11,
  ],
  Array [
    "fillRect",
    276,
    61,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    276,
    65,
    123,
    1,
  ],
  Array [
    "fillRect",
    275,
    66,
    125,
    11,
  ],
  Array [
    "fillRect",
    276,
    77,
    123,
    1,
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "click",
    278,
    75,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "arc",
    330,
    87.5,
    3.75,
    0,
    6.283185307179586,
  ],
  Array [
    "fill",
  ],
  Array [
    "set fillStyle",
    "#ededf0",
  ],
  Array [
    "fillRect",
    149,
    0,
    1,
    300,
  ],
  Array [
    "fillRect",
    0,
    15,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    63,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    79,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    95,
    365,
    1,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "fillText",
    "Marker A",
    5,
    11,
  ],
  Array [
    "fillText",
    "Dot marker E",
    5,
    43,
  ],
  Array [
    "fillText",
    "Very very very very very ver…",
    5,
    59,
  ],
  Array [
    "fillText",
    "Marker G type DOMEvent",
    5,
    75,
  ],
  Array [
    "fillText",
    "Non-interval marker F withou…",
    5,
    91,
  ],
]
`;

exports[`MarkerChart renders the normal marker chart and matches the snapshot 1`] = `
<div
  aria-labelledby="marker-chart-tab-button"
  className="markerChart"
  id="marker-chart-tab"
  role="tabpanel"
>
  <div
    className="markerSettings"
  >
    <div
      className="markerSettingsSpacer"
    />
    <div
      className="markerSettingsSearchbar"
    >
      <label
        className="markerSettingsSearchbarLabel"
      >
        Filter Markers: 
        <form
          className="idleSearchField markerSettingsSearchField"
          onSubmit={[Function]}
        >
          <input
            className="idleSearchFieldInput"
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required="required"
            title="Only display markers that match a certain name"
            type="search"
            value=""
          />
          <input
            className="idleSearchFieldButton"
            onClick={[Function]}
            tabIndex={-1}
            type="reset"
          />
        </form>
      </label>
    </div>
  </div>
  <div
    className="chartViewport"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onWheel={[Function]}
    tabIndex={0}
  >
    <div>
      <canvas
        className="chartCanvas markerChartCanvas"
        onDoubleClick={[Function]}
        onMouseDown={[Function]}
        onMouseMove={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
      />
    </div>
    <div
      className="chartViewportShiftScroll hidden"
    >
      Zoom Chart:
      <kbd
        className="chartViewportShiftScrollKbd"
      >
        Shift
      </kbd>
      <kbd
        className="chartViewportShiftScrollKbd"
      >
        Scroll
      </kbd>
    </div>
  </div>
</div>
`;

exports[`MarkerChart renders the normal marker chart and matches the snapshot 2`] = `
Array [
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    365,
    300,
  ],
  Array [
    "set lineWidth",
    1,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "measureText",
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.()< /:-_",
  ],
  Array [
    "measureText",
    "…",
  ],
  Array [
    "fillRect",
    151,
    1,
    248,
    1,
  ],
  Array [
    "fillRect",
    150,
    2,
    250,
    11,
  ],
  Array [
    "fillRect",
    151,
    13,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    151,
    17,
    248,
    1,
  ],
  Array [
    "fillRect",
    150,
    18,
    250,
    11,
  ],
  Array [
    "fillRect",
    151,
    29,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "arc",
    255,
    39.5,
    3.75,
    0,
    6.283185307179586,
  ],
  Array [
    "fill",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    276,
    49,
    248,
    1,
  ],
  Array [
    "fillRect",
    275,
    50,
    250,
    11,
  ],
  Array [
    "fillRect",
    276,
    61,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    276,
    65,
    123,
    1,
  ],
  Array [
    "fillRect",
    275,
    66,
    125,
    11,
  ],
  Array [
    "fillRect",
    276,
    77,
    123,
    1,
  ],
  Array [
    "measureText",
    "click",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "click",
    278,
    75,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "arc",
    330,
    87.5,
    3.75,
    0,
    6.283185307179586,
  ],
  Array [
    "fill",
  ],
  Array [
    "set fillStyle",
    "#ededf0",
  ],
  Array [
    "fillRect",
    149,
    0,
    1,
    300,
  ],
  Array [
    "fillRect",
    0,
    15,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    63,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    79,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    95,
    365,
    1,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "measureText",
    "Marker A",
  ],
  Array [
    "fillText",
    "Marker A",
    5,
    11,
  ],
  Array [
    "measureText",
    "Dot marker E",
  ],
  Array [
    "fillText",
    "Dot marker E",
    5,
    43,
  ],
  Array [
    "measureText",
    "Very very very very very very Very very very very very very Very very very very very very Very very very very very very Very very very very very very long Marker D",
  ],
  Array [
    "fillText",
    "Very very very very very ver…",
    5,
    59,
  ],
  Array [
    "measureText",
    "Marker G type DOMEvent",
  ],
  Array [
    "fillText",
    "Marker G type DOMEvent",
    5,
    75,
  ],
  Array [
    "measureText",
    "Non-interval marker F without data",
  ],
  Array [
    "fillText",
    "Non-interval marker F withou…",
    5,
    91,
  ],
]
`;
